File: java/var-core/src/main/java/com/oselvar/var/core/Conformance.java
package com.oselvar.var.core;

import io.cucumber.cucumberexpressions.CucumberExpressionParser;
import io.cucumber.cucumberexpressions.Node;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

/**
 * Serializes {@link Ast} nodes into the plain {@code Map}/{@code List} wire-format
 * structures that {@link CanonicalJson#canonicalStringify(Object)} turns into the
 * conformance corpus's deterministic JSON artifacts.
 *
 * <p>Port of the var-doc and registry portions of {@code var-core/src/conformance.ts}'s
 * {@code toVarDocArtifact}/{@code toRegistryArtifact} (and the equivalent
 * {@code to_var_doc_artifact}/{@code to_registry_artifact} in the Python port). Field
 * names are camelCase and must match {@code conformance/bundles/*}/golden/*.json}
 * exactly; key ordering doesn't matter here ({@link LinkedHashMap} is used purely for
 * readability while debugging) because {@link CanonicalJson} recursively sorts keys
 * itself.
 *
 * <p>Plan/trace projections are later tasks (Milestones 3-4) — this class currently
 * covers the var-doc and registry projections.
 */
public final class Conformance {

    private Conformance() {}

    /**
     * Projects a parsed {@link Ast.VarDoc} to the var-doc wire artifact: {@code
     * {path, examples, orphanAttachments}}.
     */
    public static Map<String, Object> toVarDocArtifact(Ast.VarDoc doc) {
        Map<String, Object> out = new LinkedHashMap<>();
        out.put("path", doc.path());
        out.put("examples", doc.examples().stream().map(Conformance::example).toList());
        out.put(
                "orphanAttachments",
                doc.orphanAttachments().stream().map(Conformance::tableOrFence).toList());
        return out;
    }

    /**
     * Projects a {@link Registry} to the registry wire artifact: {@code {steps:
     * [{expression, parameterTypeNames}], parameterTypes: [{name, regexp}]}}.
     *
     * <p>Port of {@code toRegistryArtifact} in {@code conformance.ts} (and
     * {@code to_registry_artifact} in the Python port). No conformance bundle
     * currently exercises {@code defineParameterType} (every {@code golden/
     * registry.json}'s {@code parameterTypes} is {@code []} — see the plan's
     * deferred list), so unlike TS/Python this overload takes no explicit custom-
     * parameter-types argument; add one if/when a bundle needs it.
     */
    public static Map<String, Object> toRegistryArtifact(Registry registry) {
        Map<String, Object> out = new LinkedHashMap<>();
        out.put("steps", registry.steps().stream().map(Conformance::step).toList());
        out.put("parameterTypes", List.of());
        return out;
    }

    private static Map<String, Object> step(Registry.StepRegistration s) {
        Map<String, Object> out = new LinkedHashMap<>();
        out.put("expression", s.expression());
        out.put("parameterTypeNames", parameterTypeNames(s.expression()));
        return out;
    }

    /**
     * Parameter-type names in source order, read from the expression's parsed AST
     * (authoritative). A naive {@code {...}} regex miscounts on escaped braces
     * ({@code \{}/{@code \}}), which are literal text, not parameters. Cucumber
     * rejects parameters inside optionals/alternation, so they only appear at the
     * top level, but this recurses defensively, mirroring {@code conformance.ts}'s
     * {@code parameterTypeNames}.
     *
     * <p>Java's {@code CucumberExpression} (unlike the TS/Python libraries) does not
     * expose its parsed AST or parameter-type list publicly — confirmed via {@code
     * javap -p}, no {@code getAst()}/{@code getParameterTypes()} escape the class.
     * This re-parses {@code source} with the library's own public {@link
     * CucumberExpressionParser#parse(String)}, which is exactly what {@code
     * CucumberExpression}'s constructor does internally to build its regex —
     * reproducing an identical {@link Node} tree, empirically confirmed by dumping
     * it for an expression exercising nested parameters and an escaped brace (each
     * {@code PARAMETER_NODE} has exactly one {@code TEXT_NODE} child holding the
     * name, and {@code \{escaped\}} parses as a single literal {@code TEXT_NODE},
     * never a parameter). {@link Node#text()} recurses the same way internally but
     * is package-private; {@link #nodeText} reimplements it against the class's
     * public surface ({@link Node#token()}/{@link Node#nodes()}) instead of
     * reflecting into the library.
     */
    static List<String> parameterTypeNames(String source) {
        Node ast = new CucumberExpressionParser().parse(source);
        List<String> names = new ArrayList<>();
        collectParameterNames(ast, names);
        return names;
    }

    private static void collectParameterNames(Node node, List<String> names) {
        if (node.type() == Node.Type.PARAMETER_NODE) {
            names.add(nodeText(node));
            return;
        }
        List<Node> children = node.nodes();
        if (children != null) {
            for (Node child : children) collectParameterNames(child, names);
        }
    }

    /** Reimplements {@code Node#text()} (package-private in the library) publicly. */
    private static String nodeText(Node node) {
        String token = node.token();
        if (token != null) return token;
        StringBuilder sb = new StringBuilder();
        List<Node> children = node.nodes();
        if (children != null) {
            for (Node child : children) sb.append(nodeText(child));
        }
        return sb.toString();
    }

    /** Dispatches on the sealed {@link Ast.TableOrFence} union (the orphan-attachment type). */
    private static Map<String, Object> tableOrFence(Ast.TableOrFence tableOrFence) {
        return switch (tableOrFence) {
            case Ast.Table t -> table(t);
            case Ast.Fence f -> fence(f);
        };
    }

    private static Map<String, Object> example(Ast.Example example) {
        Map<String, Object> out = new LinkedHashMap<>();
        out.put("scopeStack", List.copyOf(example.scopeStack()));
        out.put("span", span(example.span()));
        out.put("body", example.body().stream().map(Conformance::block).toList());
        return out;
    }

    /** Dispatches on the sealed {@link Ast.Block} union — exhaustive, no default branch. */
    private static Map<String, Object> block(Ast.Block block) {
        return switch (block) {
            case Ast.Heading h -> heading(h);
            case Ast.Paragraph p -> paragraph(p);
            case Ast.ListItem l -> listItem(l);
            case Ast.Blockquote b -> blockquote(b);
            case Ast.Table t -> table(t);
            case Ast.Fence f -> fence(f);
            case Ast.ThematicBreak t -> thematicBreak(t);
        };
    }

    private static Map<String, Object> heading(Ast.Heading h) {
        Map<String, Object> out = new LinkedHashMap<>();
        out.put("kind", "heading");
        out.put("level", h.level());
        out.put("text", h.text());
        out.put("span", span(h.span()));
        return out;
    }

    private static Map<String, Object> paragraph(Ast.Paragraph p) {
        Map<String, Object> out = new LinkedHashMap<>();
        out.put("kind", "paragraph");
        out.put("text", p.text());
        out.put("span", span(p.span()));
        out.put("inlineMap", inlineMap(p.inlineMap()));
        return out;
    }

    private static Map<String, Object> listItem(Ast.ListItem l) {
        Map<String, Object> out = new LinkedHashMap<>();
        out.put("kind", "list_item");
        out.put("text", l.text());
        out.put("span", span(l.span()));
        out.put("inlineMap", inlineMap(l.inlineMap()));
        out.put("ordered", l.ordered());
        out.put("markerSpan", span(l.markerSpan()));
        return out;
    }

    private static Map<String, Object> blockquote(Ast.Blockquote b) {
        Map<String, Object> out = new LinkedHashMap<>();
        out.put("kind", "blockquote");
        out.put("text", b.text());
        out.put("span", span(b.span()));
        out.put("inlineMap", inlineMap(b.inlineMap()));
        return out;
    }

    private static Map<String, Object> table(Ast.Table t) {
        Map<String, Object> out = new LinkedHashMap<>();
        out.put("kind", "table");
        out.put("span", span(t.span()));
        out.put("header", row(t.header()));
        out.put("rows", t.rows().stream().map(Conformance::row).toList());
        return out;
    }

    private static Map<String, Object> row(Ast.Row r) {
        Map<String, Object> out = new LinkedHashMap<>();
        out.put("cells", List.copyOf(r.cells()));
        out.put("cellSpans", r.cellSpans().stream().map(Conformance::span).toList());
        out.put("span", span(r.span()));
        return out;
    }

    private static Map<String, Object> fence(Ast.Fence f) {
        Map<String, Object> out = new LinkedHashMap<>();
        out.put("kind", "fence");
        out.put("span", span(f.span()));
        out.put("info", f.info());
        out.put("body", f.body());
        out.put("bodySpan", span(f.bodySpan()));
        return out;
    }

    private static Map<String, Object> thematicBreak(Ast.ThematicBreak t) {
        Map<String, Object> out = new LinkedHashMap<>();
        out.put("kind", "thematic_break");
        out.put("span", span(t.span()));
        return out;
    }

    private static List<Object> inlineMap(List<Ast.InlineOffset> inlineMap) {
        return inlineMap.stream().<Object>map(Conformance::inlineOffset).toList();
    }

    private static Map<String, Object> inlineOffset(Ast.InlineOffset o) {
        Map<String, Object> out = new LinkedHashMap<>();
        out.put("textOffset", o.textOffset());
        out.put("sourceOffset", o.sourceOffset());
        return out;
    }

    private static Map<String, Object> span(Span s) {
        Map<String, Object> out = new LinkedHashMap<>();
        out.put("startOffset", s.startOffset());
        out.put("endOffset", s.endOffset());
        out.put("startLine", s.startLine());
        out.put("startCol", s.startCol());
        out.put("endLine", s.endLine());
        out.put("endCol", s.endCol());
        return out;
    }
}

File: typescript/packages/core/src/conformance.ts
import { type CucumberExpression, type Node, NodeType } from '@cucumber/cucumber-expressions'
import type { Fence, Table, VarDoc } from './ast.ts'
import { isCellMismatchError, ReturnShapeError } from './cell-diff.ts'
import type { DiagnosticCode, Severity } from './diagnostics.ts'
import { isDocStringMismatchError } from './doc-string-diff.ts'
import { collectExamples, isUnexpectedPassError, type StepObservation } from './execute.ts'
import { failureAnchor } from './failure-anchor.ts'
import { plan as buildPlan, type ExecutionPlan } from './plan.ts'
import type { Registry } from './registry.ts'
import type { Span } from './span.ts'

// ---- Artifact types (the serialized contract) -----------------------------

export type VarDocArtifact = {
  readonly path: string
  readonly examples: VarDoc['examples']
  readonly orphanAttachments: ReadonlyArray<Table | Fence>
}

export type RegistryArtifact = {
  readonly steps: ReadonlyArray<{
    readonly expression: string
    readonly parameterTypeNames: ReadonlyArray<string>
  }>
  // Custom parameter types (name + source regexp). Empty until a bundle uses
  // defineParameterType — see the plan's deferred list.
  readonly parameterTypes: ReadonlyArray<{ readonly name: string; readonly regexp: string }>
}

export type PlanArtifact = {
  readonly examples: ReadonlyArray<{
    readonly name: string
    readonly scopeStack: ReadonlyArray<string>
    readonly span: Span
    readonly expectedOutcome: 'pass' | 'fail'
    readonly expectedErrorMessage?: string
    readonly steps: ReadonlyArray<{
      readonly text: string
      readonly matchSpan: Span
      readonly paramSpans: ReadonlyArray<Span>
      readonly matchedExpression: string
      readonly args: ReadonlyArray<{
        readonly value: string
        readonly parameterType: string | null
      }>
      readonly dataTable?: Table
      readonly docString?: {
        readonly content: string
        readonly contentType: string
        readonly span: Span
      }
    }>
  }>
  readonly diagnostics: ReadonlyArray<{
    readonly code: DiagnosticCode
    readonly severity: Severity
    readonly span: Span
  }>
}

// `message` is present on every core-generated kind and absent on `thrown`; see
// toFailureArtifact for why.
export type FailureArtifact =
  | {
      readonly kind: 'cell-mismatch'
      readonly line: number
      readonly anchor: Span
      readonly message: string
      readonly cells: ReadonlyArray<{
        readonly column: string
        readonly expected: string
        readonly actual: string
        readonly span: Span
      }>
    }
  | {
      readonly kind: 'doc-string-mismatch'
      readonly line: number
      readonly anchor: Span
      readonly message: string
      readonly diff: { readonly expected: string; readonly actual: string; readonly span: Span }
    }
  | {
      readonly kind: 'return-shape'
      readonly line: number
      readonly anchor: Span
      readonly message: string
    }
  | { readonly kind: 'thrown'; readonly line: number; readonly anchor: Span }
  | {
      readonly kind: 'unexpected-pass'
      readonly line: number
      readonly anchor: Span
      readonly message: string
    }

export type StepTrace = {
  readonly exampleName: string
  readonly ordinal: number
  readonly stepText: string
  readonly matchedExpression: string
  readonly contextKey: { readonly exampleName: string; readonly stepFile: string }
  readonly outcome: 'pass' | 'fail' | 'skipped'
  readonly failure?: FailureArtifact
}

export type TraceArtifact = {
  readonly examples: ReadonlyArray<{
    readonly name: string
    readonly outcome: 'pass' | 'fail'
    readonly steps: ReadonlyArray<StepTrace>
  }>
}

export type BundleArtifacts = {
  readonly varDoc: VarDocArtifact
  readonly registry: RegistryArtifact
  readonly plan: PlanArtifact
  readonly trace: TraceArtifact
}

// ---- Canonical serialization ----------------------------------------------

function sortKeys(value: unknown): unknown {
  if (Array.isArray(value)) return value.map(sortKeys)
  if (value && typeof value === 'object') {
    const out: Record<string, unknown> = {}
    for (const key of Object.keys(value as Record<string, unknown>).sort()) {
      out[key] = sortKeys((value as Record<string, unknown>)[key])
    }
    return out
  }
  return value
}

// Deterministic JSON: recursively key-sorted, 2-space indent, LF endings,
// trailing newline. The wire format every implementation must reproduce.
export function canonicalStringify(value: unknown): string {
  return `${JSON.stringify(sortKeys(value), null, 2)}\n`
}

// `path/to/foo.steps.ts` -> `foo.steps` ; `s.ts` -> `s`. Normalizes step-def
// file references so TS and Python fixtures serialize identically. Internal
// (not exported) — used only within this module.
function fileStem(path: string): string {
  const base = path.split(/[\\/]/).pop() ?? path
  return base.replace(/\.[^.]+$/, '')
}

// Parameter-type names in source order, read from the compiled expression's
// AST (authoritative). A naive `{...}` regex miscounts on escaped braces
// (`\{`/`\}`), which are literal text, not parameters. Cucumber rejects
// parameters inside optionals/alternation, so they only appear at the top
// level, but we recurse defensively. Internal — used only within this module.
function parameterTypeNames(compiled: CucumberExpression): ReadonlyArray<string> {
  const names: string[] = []
  const visit = (node: Node): void => {
    if (node.type === NodeType.parameter) {
      names.push(node.text())
      return
    }
    for (const child of node.nodes ?? []) visit(child)
  }
  visit(compiled.ast)
  return names
}

export function toVarDocArtifact(doc: VarDoc): VarDocArtifact {
  return { path: doc.path, examples: doc.examples, orphanAttachments: doc.orphanAttachments }
}

export function toRegistryArtifact(
  registry: Registry,
  parameterTypes: ReadonlyArray<{ name: string; regexp: string }> = [],
): RegistryArtifact {
  return {
    steps: registry.steps.map((s) => ({
      expression: s.expression,
      parameterTypeNames: parameterTypeNames(s.compiled),
    })),
    parameterTypes: parameterTypes.map((p) => ({ name: p.name, regexp: p.regexp })),
  }
}

export function toPlanArtifact(plan: ExecutionPlan): PlanArtifact {
  return {
    examples: plan.examples.map((ex) => ({
      name: ex.name,
      scopeStack: ex.scopeStack,
      span: ex.span,
      expectedOutcome: ex.expectedOutcome ?? 'pass',
      ...(ex.expectedErrorMessage ? { expectedErrorMessage: ex.expectedErrorMessage } : {}),
      steps: ex.steps.map((step) => {
        const stepNames = parameterTypeNames(step.stepDef.compiled)
        return {
          text: step.text,
          matchSpan: step.matchSpan,
          paramSpans: step.paramSpans,
          matchedExpression: step.stepDef.expression,
          args: step.paramSpans.map((span, i) => ({
            value: plan.varDoc.source.slice(span.startOffset, span.endOffset),
            parameterType: stepNames[i] ?? null,
          })),
          ...(step.dataTable ? { dataTable: step.dataTable } : {}),
          ...(step.docString ? { docString: step.docString } : {}),
        }
      }),
    })),
    diagnostics: plan.diagnostics.map((d) => ({
      code: d.code,
      severity: d.severity,
      span: d.span,
    })),
  }
}

// `line` is the failing step's own 1-based line in the `.md`
// (`matchSpan.startLine`) and `anchor` is where the failure points — the
// failureAnchor rule (first failing cell span / doc string body / the step's
// match span). Both are deterministic, language-agnostic source positions, not
// values scraped from a JS stack trace, so every port reproduces them; pinning
// `anchor` in the goldens is what keeps each port's editor-facing failure
// location (stack augmentation) from drifting.
export function toFailureArtifact(error: unknown, matchSpan: Span): FailureArtifact {
  const line = matchSpan.startLine
  const anchor = failureAnchor(error, matchSpan)
  // `message` is projected for the kinds the CORE generates, so the corpus pins
  // their wording across ports — an `error` fence matches by substring, so a port
  // that words or quotes a message differently fails a spec its siblings pass.
  // `thrown` is deliberately excluded: those messages come from user handlers and
  // from language runtimes, which differ legitimately (dividing by zero reads
  // "division by zero" in Python, "/ by zero" in Java, "runtime error: integer
  // divide by zero" in Go).
  if (isCellMismatchError(error)) {
    return {
      kind: 'cell-mismatch',
      line,
      anchor,
      message: error.message,
      cells: error.cells
        .filter((c) => !c.ok)
        .map((c) => ({ column: c.column, expected: c.expected, actual: c.actual, span: c.span })),
    }
  }
  if (isDocStringMismatchError(error)) {
    return {
      kind: 'doc-string-mismatch',
      line,
      anchor,
      message: error.message,
      diff: { expected: error.diff.expected, actual: error.diff.actual, span: error.diff.span },
    }
  }
  if (error instanceof ReturnShapeError)
    return { kind: 'return-shape', line, anchor, message: error.message }
  if (isUnexpectedPassError(error))
    return { kind: 'unexpected-pass', line, anchor, message: error.message }
  return { kind: 'thrown', line, anchor }
}

export async function runConformance(
  varDoc: VarDoc,
  registry: Registry,
  createContext: (stepFile: string) => unknown | Promise<unknown>,
  parameterTypes: ReadonlyArray<{ name: string; regexp: string }> = [],
): Promise<BundleArtifacts> {
  const execution = buildPlan(varDoc, registry)

  const observed = new Map<number, StepObservation[]>()
  const queue = collectExamples(execution, {
    reporter: { diagnostic: () => {} }, // diagnostics are captured in the plan artifact
    createContext,
    observer: {
      step: (o) => {
        const list = observed.get(o.exampleIndex) ?? []
        list.push(o)
        observed.set(o.exampleIndex, list)
      },
    },
  })

  const traceExamples = []
  for (let k = 0; k < queue.length; k++) {
    const { name, run } = queue[k] as { name: string; run: () => void | Promise<void> }
    let outcome: 'pass' | 'fail' = 'pass'
    try {
      await run()
    } catch {
      outcome = 'fail'
    }
    const planned = execution.examples[k]
    const obs = observed.get(k) ?? []
    const steps: StepTrace[] = (planned?.steps ?? []).map((step, i) => {
      const matches = obs.filter((x) => x.ordinal === i + 1)
      const o = matches.find((m) => m.outcome === 'fail') ?? matches[matches.length - 1]
      const stepOutcome = o ? o.outcome : 'skipped'
      return {
        exampleName: name,
        ordinal: i + 1,
        stepText: step.text,
        matchedExpression: step.stepDef.expression,
        contextKey: { exampleName: name, stepFile: fileStem(step.stepDef.expressionSourceFile) },
        outcome: stepOutcome,
        ...(stepOutcome === 'fail' ? { failure: toFailureArtifact(o?.error, step.matchSpan) } : {}),
      }
    })
    traceExamples.push({ name, outcome, steps })
  }

  return {
    varDoc: toVarDocArtifact(varDoc),
    registry: toRegistryArtifact(registry, parameterTypes),
    plan: toPlanArtifact(execution),
    trace: { examples: traceExamples },
  }
}
